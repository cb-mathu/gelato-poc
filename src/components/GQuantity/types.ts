import { VueComponentProps } from '../../types';
import { ExtractPropTypes, InputHTMLAttributes } from 'vue';
import { QUANTITY_PROPS } from './constants';

export type QuantityProps = Omit<ExtractPropTypes<typeof QUANTITY_PROPS>, 'errorMessage'>;
export type NativeInputProps = Omit<InputHTMLAttributes, keyof QuantityProps>;
export type MergedQuantityProps = Partial<QuantityProps & VueComponentProps & NativeInputProps>;
