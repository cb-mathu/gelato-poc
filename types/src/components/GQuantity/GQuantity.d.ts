import { MergedQuantityProps } from './types';
declare const _default: new () => {
    $props: MergedQuantityProps;
};
export default _default;
