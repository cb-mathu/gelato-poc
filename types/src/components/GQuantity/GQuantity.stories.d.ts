import { QuantityProps } from './types';
declare const _default: {
    title: string;
    component: new () => {
        $props: Partial<QuantityProps & {
            onCopy?: ((payload: ClipboardEvent) => void) | undefined;
            onCut?: ((payload: ClipboardEvent) => void) | undefined;
            onPaste?: ((payload: ClipboardEvent) => void) | undefined;
            onCompositionend?: ((payload: CompositionEvent) => void) | undefined;
            onCompositionstart?: ((payload: CompositionEvent) => void) | undefined;
            onCompositionupdate?: ((payload: CompositionEvent) => void) | undefined;
            onDrag?: ((payload: DragEvent) => void) | undefined;
            onDragend?: ((payload: DragEvent) => void) | undefined;
            onDragenter?: ((payload: DragEvent) => void) | undefined;
            onDragexit?: ((payload: DragEvent) => void) | undefined;
            onDragleave?: ((payload: DragEvent) => void) | undefined;
            onDragover?: ((payload: DragEvent) => void) | undefined;
            onDragstart?: ((payload: DragEvent) => void) | undefined;
            onDrop?: ((payload: DragEvent) => void) | undefined;
            onFocus?: ((payload: FocusEvent) => void) | undefined;
            onFocusin?: ((payload: FocusEvent) => void) | undefined;
            onFocusout?: ((payload: FocusEvent) => void) | undefined;
            onBlur?: ((payload: FocusEvent) => void) | undefined;
            onChange?: ((payload: Event) => void) | undefined;
            onBeforeinput?: ((payload: Event) => void) | undefined;
            onInput?: ((payload: Event) => void) | undefined;
            onReset?: ((payload: Event) => void) | undefined;
            onSubmit?: ((payload: Event) => void) | undefined;
            onInvalid?: ((payload: Event) => void) | undefined;
            onLoad?: ((payload: Event) => void) | undefined;
            onError?: ((payload: Event) => void) | undefined;
            onKeydown?: ((payload: KeyboardEvent) => void) | undefined;
            onKeypress?: ((payload: KeyboardEvent) => void) | undefined;
            onKeyup?: ((payload: KeyboardEvent) => void) | undefined;
            onAuxclick?: ((payload: MouseEvent) => void) | undefined;
            onClick?: ((payload: MouseEvent) => void) | undefined;
            onContextmenu?: ((payload: MouseEvent) => void) | undefined;
            onDblclick?: ((payload: MouseEvent) => void) | undefined;
            onMousedown?: ((payload: MouseEvent) => void) | undefined;
            onMouseenter?: ((payload: MouseEvent) => void) | undefined;
            onMouseleave?: ((payload: MouseEvent) => void) | undefined;
            onMousemove?: ((payload: MouseEvent) => void) | undefined;
            onMouseout?: ((payload: MouseEvent) => void) | undefined;
            onMouseover?: ((payload: MouseEvent) => void) | undefined;
            onMouseup?: ((payload: MouseEvent) => void) | undefined;
            onAbort?: ((payload: Event) => void) | undefined;
            onCanplay?: ((payload: Event) => void) | undefined;
            onCanplaythrough?: ((payload: Event) => void) | undefined;
            onDurationchange?: ((payload: Event) => void) | undefined;
            onEmptied?: ((payload: Event) => void) | undefined;
            onEncrypted?: ((payload: Event) => void) | undefined;
            onEnded?: ((payload: Event) => void) | undefined;
            onLoadeddata?: ((payload: Event) => void) | undefined;
            onLoadedmetadata?: ((payload: Event) => void) | undefined;
            onLoadstart?: ((payload: Event) => void) | undefined;
            onPause?: ((payload: Event) => void) | undefined;
            onPlay?: ((payload: Event) => void) | undefined;
            onPlaying?: ((payload: Event) => void) | undefined;
            onProgress?: ((payload: Event) => void) | undefined;
            onRatechange?: ((payload: Event) => void) | undefined;
            onSeeked?: ((payload: Event) => void) | undefined;
            onSeeking?: ((payload: Event) => void) | undefined;
            onStalled?: ((payload: Event) => void) | undefined;
            onSuspend?: ((payload: Event) => void) | undefined;
            onTimeupdate?: ((payload: Event) => void) | undefined;
            onVolumechange?: ((payload: Event) => void) | undefined;
            onWaiting?: ((payload: Event) => void) | undefined;
            onSelect?: ((payload: Event) => void) | undefined;
            onScroll?: ((payload: UIEvent) => void) | undefined;
            onTouchcancel?: ((payload: TouchEvent) => void) | undefined;
            onTouchend?: ((payload: TouchEvent) => void) | undefined;
            onTouchmove?: ((payload: TouchEvent) => void) | undefined;
            onTouchstart?: ((payload: TouchEvent) => void) | undefined;
            onPointerdown?: ((payload: PointerEvent) => void) | undefined;
            onPointermove?: ((payload: PointerEvent) => void) | undefined;
            onPointerup?: ((payload: PointerEvent) => void) | undefined;
            onPointercancel?: ((payload: PointerEvent) => void) | undefined;
            onPointerenter?: ((payload: PointerEvent) => void) | undefined;
            onPointerleave?: ((payload: PointerEvent) => void) | undefined;
            onPointerover?: ((payload: PointerEvent) => void) | undefined;
            onPointerout?: ((payload: PointerEvent) => void) | undefined;
            onWheel?: ((payload: WheelEvent) => void) | undefined;
            onAnimationstart?: ((payload: AnimationEvent) => void) | undefined;
            onAnimationend?: ((payload: AnimationEvent) => void) | undefined;
            onAnimationiteration?: ((payload: AnimationEvent) => void) | undefined;
            onTransitionend?: ((payload: TransitionEvent) => void) | undefined;
            onTransitionstart?: ((payload: TransitionEvent) => void) | undefined;
        } & import("vue").AllowedComponentProps & import("vue").VNodeProps & {
            class: string;
        } & {
            id?: string | undefined;
        } & import("./types").NativeInputProps>;
    };
    argTypes: {
        min: {
            description: string;
            control: {
                type: string;
            };
        };
        max: {
            description: string;
            control: {
                type: string;
            };
        };
        step: {
            description: string;
            control: {
                type: string;
            };
        };
    };
};
export default _default;
export declare const Story: (args: QuantityProps) => {
    setup(): () => JSX.Element;
};
export declare const DefaultQuantity: (props?: Partial<QuantityProps & {
    onCopy?: ((payload: ClipboardEvent) => void) | undefined;
    onCut?: ((payload: ClipboardEvent) => void) | undefined;
    onPaste?: ((payload: ClipboardEvent) => void) | undefined;
    onCompositionend?: ((payload: CompositionEvent) => void) | undefined;
    onCompositionstart?: ((payload: CompositionEvent) => void) | undefined;
    onCompositionupdate?: ((payload: CompositionEvent) => void) | undefined;
    onDrag?: ((payload: DragEvent) => void) | undefined;
    onDragend?: ((payload: DragEvent) => void) | undefined;
    onDragenter?: ((payload: DragEvent) => void) | undefined;
    onDragexit?: ((payload: DragEvent) => void) | undefined;
    onDragleave?: ((payload: DragEvent) => void) | undefined;
    onDragover?: ((payload: DragEvent) => void) | undefined;
    onDragstart?: ((payload: DragEvent) => void) | undefined;
    onDrop?: ((payload: DragEvent) => void) | undefined;
    onFocus?: ((payload: FocusEvent) => void) | undefined;
    onFocusin?: ((payload: FocusEvent) => void) | undefined;
    onFocusout?: ((payload: FocusEvent) => void) | undefined;
    onBlur?: ((payload: FocusEvent) => void) | undefined;
    onChange?: ((payload: Event) => void) | undefined;
    onBeforeinput?: ((payload: Event) => void) | undefined;
    onInput?: ((payload: Event) => void) | undefined;
    onReset?: ((payload: Event) => void) | undefined;
    onSubmit?: ((payload: Event) => void) | undefined;
    onInvalid?: ((payload: Event) => void) | undefined;
    onLoad?: ((payload: Event) => void) | undefined;
    onError?: ((payload: Event) => void) | undefined;
    onKeydown?: ((payload: KeyboardEvent) => void) | undefined;
    onKeypress?: ((payload: KeyboardEvent) => void) | undefined;
    onKeyup?: ((payload: KeyboardEvent) => void) | undefined;
    onAuxclick?: ((payload: MouseEvent) => void) | undefined;
    onClick?: ((payload: MouseEvent) => void) | undefined;
    onContextmenu?: ((payload: MouseEvent) => void) | undefined;
    onDblclick?: ((payload: MouseEvent) => void) | undefined;
    onMousedown?: ((payload: MouseEvent) => void) | undefined;
    onMouseenter?: ((payload: MouseEvent) => void) | undefined;
    onMouseleave?: ((payload: MouseEvent) => void) | undefined;
    onMousemove?: ((payload: MouseEvent) => void) | undefined;
    onMouseout?: ((payload: MouseEvent) => void) | undefined;
    onMouseover?: ((payload: MouseEvent) => void) | undefined;
    onMouseup?: ((payload: MouseEvent) => void) | undefined;
    onAbort?: ((payload: Event) => void) | undefined;
    onCanplay?: ((payload: Event) => void) | undefined;
    onCanplaythrough?: ((payload: Event) => void) | undefined;
    onDurationchange?: ((payload: Event) => void) | undefined;
    onEmptied?: ((payload: Event) => void) | undefined;
    onEncrypted?: ((payload: Event) => void) | undefined;
    onEnded?: ((payload: Event) => void) | undefined;
    onLoadeddata?: ((payload: Event) => void) | undefined;
    onLoadedmetadata?: ((payload: Event) => void) | undefined;
    onLoadstart?: ((payload: Event) => void) | undefined;
    onPause?: ((payload: Event) => void) | undefined;
    onPlay?: ((payload: Event) => void) | undefined;
    onPlaying?: ((payload: Event) => void) | undefined;
    onProgress?: ((payload: Event) => void) | undefined;
    onRatechange?: ((payload: Event) => void) | undefined;
    onSeeked?: ((payload: Event) => void) | undefined;
    onSeeking?: ((payload: Event) => void) | undefined;
    onStalled?: ((payload: Event) => void) | undefined;
    onSuspend?: ((payload: Event) => void) | undefined;
    onTimeupdate?: ((payload: Event) => void) | undefined;
    onVolumechange?: ((payload: Event) => void) | undefined;
    onWaiting?: ((payload: Event) => void) | undefined;
    onSelect?: ((payload: Event) => void) | undefined;
    onScroll?: ((payload: UIEvent) => void) | undefined;
    onTouchcancel?: ((payload: TouchEvent) => void) | undefined;
    onTouchend?: ((payload: TouchEvent) => void) | undefined;
    onTouchmove?: ((payload: TouchEvent) => void) | undefined;
    onTouchstart?: ((payload: TouchEvent) => void) | undefined;
    onPointerdown?: ((payload: PointerEvent) => void) | undefined;
    onPointermove?: ((payload: PointerEvent) => void) | undefined;
    onPointerup?: ((payload: PointerEvent) => void) | undefined;
    onPointercancel?: ((payload: PointerEvent) => void) | undefined;
    onPointerenter?: ((payload: PointerEvent) => void) | undefined;
    onPointerleave?: ((payload: PointerEvent) => void) | undefined;
    onPointerover?: ((payload: PointerEvent) => void) | undefined;
    onPointerout?: ((payload: PointerEvent) => void) | undefined;
    onWheel?: ((payload: WheelEvent) => void) | undefined;
    onAnimationstart?: ((payload: AnimationEvent) => void) | undefined;
    onAnimationend?: ((payload: AnimationEvent) => void) | undefined;
    onAnimationiteration?: ((payload: AnimationEvent) => void) | undefined;
    onTransitionend?: ((payload: TransitionEvent) => void) | undefined;
    onTransitionstart?: ((payload: TransitionEvent) => void) | undefined;
} & import("vue").AllowedComponentProps & import("vue").VNodeProps & {
    class: string;
} & {
    id?: string | undefined;
} & import("./types").NativeInputProps> | undefined) => JSX.Element;
export declare const DisabledQuantity: () => JSX.Element;
export declare const ReadOnlyQuantity: () => JSX.Element;
export declare const InvalidQuantity: () => JSX.Element;
