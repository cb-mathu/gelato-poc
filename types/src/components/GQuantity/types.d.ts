import { VueComponentProps } from '../../types';
import { ExtractPropTypes, InputHTMLAttributes } from 'vue';
import { QUANTITY_PROPS } from './constants';
export declare type QuantityProps = Omit<ExtractPropTypes<typeof QUANTITY_PROPS>, 'errorMessage'>;
export declare type NativeInputProps = Omit<InputHTMLAttributes, keyof QuantityProps>;
export declare type MergedQuantityProps = Partial<QuantityProps & VueComponentProps & NativeInputProps>;
