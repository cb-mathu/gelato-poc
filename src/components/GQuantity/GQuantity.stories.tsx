import GQuantity from './GQuantity';
import { ref } from 'vue';
import { GNotification } from '../GNotification';
import { MergedQuantityProps, QuantityProps } from './types';

export default {
  title: 'Component/Forms/GQuantity',
  component: GQuantity,
  argTypes: {
    min: {
      description: 'Max value',
      control: {
        type: 'number',
      },
    },
    max: {
      description: 'Min value',
      control: {
        type: 'number',
      },
    },
    step: {
      description: 'Stepper value',
      control: {
        type: 'number',
      },
    },
  },
};

const Template = (args: QuantityProps) => ({
  setup() {
    const input = ref(6);
    return () => (
      <div style="display:grid;grid-gap: 0.5rem;">
        <div>
          <GQuantity {...args} v-model={input.value} />
        </div>
        <GNotification style="margin-top: 1rem">
          Entered Value:<b>{input.value}</b>
        </GNotification>
      </div>
    );
  },
});

export const Story = Template.bind({});

export const DefaultQuantity = (props?: MergedQuantityProps) => <GQuantity {...(props as any)} />;

export const DisabledQuantity = () => <GQuantity value={5} disabled />;

export const ReadOnlyQuantity = () => <GQuantity value={5} readonly />;

export const InvalidQuantity = () => <GQuantity value={5} invalid />;
