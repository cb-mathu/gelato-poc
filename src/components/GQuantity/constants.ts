import { PropType } from 'vue';
import { BASE_FORM_PROPS } from '../../utils/props';

export const QUANTITY_PROPS = {
  ...BASE_FORM_PROPS,
  'onUpdate:modelValue': Function as PropType<(value: number) => void>,
};
