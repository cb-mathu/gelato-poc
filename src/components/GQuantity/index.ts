import GQuantity from './GQuantity';

export { GQuantity };
