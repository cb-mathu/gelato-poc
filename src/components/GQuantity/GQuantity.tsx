import { defineComponent } from 'vue';
import { useId, useInput, FieldTypes } from '../../composables';
import { QUANTITY_PROPS } from './constants';
import { Input, InputTypes } from '../core/Input';
import { Minus, Plus } from '../icons';
import { QuantityProps, NativeInputProps, MergedQuantityProps } from './types';
import { Clickable } from '../../utils/accessibility';

const GQuantity = defineComponent({
  name: 'GQuantity',
  props: QUANTITY_PROPS,
  emits: ['update:modelValue', 'blur'],

  setup(props: QuantityProps, { attrs }: { attrs: NativeInputProps }) {
    const { internalValue, ...listeners } = useInput({
      props,
    });

    return () => {
      const onIncrement = (step: number) => {
        const _value = parseInt(internalValue.value, 10);
        if (attrs.max !== undefined && _value >= attrs.max) {
          return;
        }
        internalValue.value = _value + step;
      };

      const onDecrement = (step: number) => {
        const _value = parseInt(internalValue.value, 10);
        if ((attrs.min !== undefined && _value <= attrs.min) || _value <= 0) {
          return;
        }
        internalValue.value = _value - step;
      };

      const propsWeControl: any = {
        id: props.id || useId('GQuantity'),
        type: FieldTypes.Number,
        ...listeners,
      };

      const attributes = {
        ...attrs,
        ...props,
        ...propsWeControl,
      };

      const { max = Infinity, min = -Infinity, step = 1 } = attributes;

      const classes = {
        [`g-quantity`]: true,
        [`g-quantity--invalid`]: props.invalid,
        [`g-quantity--readonly`]: props.readonly,
        [`g-quantity--disabled`]: props.disabled,
      };

      const ariaAttributes = {
        ['aria-live']: 'assertive',
        ['aria-label']: 'units',
      };

      const disableControl = (condition: boolean) => {
        const isDisabled = props.readonly || props.disabled;
        return condition || isDisabled;
      };

      const isDecDisabled = disableControl(internalValue.value <= min || internalValue.value - step < min);
      const isIncDisabled = disableControl(internalValue.value >= max || internalValue.value + step > max);

      return (
        <div class={classes}>
          <Clickable
            as="button"
            role="spinbutton"
            disabled={isDecDisabled}
            class="g-quantity__change g-quantity__decrement"
            onClick={() => !isDecDisabled && onDecrement(step)}
          >
            <Minus />
          </Clickable>
          <Input
            {...attributes}
            {...ariaAttributes}
            modelValue={internalValue.value}
            class="g-quantity__element"
            type={InputTypes.Number}
          />
          <Clickable
            as="button"
            role="spinbutton"
            disabled={isIncDisabled}
            class="g-quantity__change g-quantity__increment"
            onClick={() => !isIncDisabled && onIncrement(step)}
          >
            <Plus />
          </Clickable>
        </div>
      );
    };
  },
});

export default GQuantity as {
  new (): {
    $props: MergedQuantityProps;
  };
};
