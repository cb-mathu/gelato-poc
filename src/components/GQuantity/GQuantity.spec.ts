import { render, screen, fireEvent } from '@testing-library/vue';
import userEvent from '@testing-library/user-event';
import * as stories from './GQuantity.stories';

const { DefaultQuantity, DisabledQuantity, InvalidQuantity, ReadOnlyQuantity } = stories;

const fieldName = 'plan-qty';
const fieldValue = '6';

const fireIncControl = () => {
  const [, incElement] = screen.getAllByRole('button');
  fireEvent.click(incElement);
};

const fireDecControl = () => {
  const [decElement] = screen.getAllByRole('button');
  fireEvent.click(decElement);
};

const fireControls = () => {
  fireIncControl();
  fireDecControl();
};

describe('GQuantity : ', () => {
  test('renders quantity with default args', () => {
    const handleClick = jest.fn();
    const { container } = render(DefaultQuantity({ onClick: handleClick }));
    fireControls();
    expect(handleClick).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  test('renders quantity with native attributes', () => {
    const { container } = render(DefaultQuantity({ min: 4, max: 5, step: 2 }));
    expect(container).toMatchSnapshot();
  });

  it('pressing tab should focus elements', () => {
    const { container } = render(DefaultQuantity());
    userEvent.tab();
    const [dec, inc] = screen.getAllByRole('button');
    expect(dec).toHaveFocus();
    userEvent.tab();
    expect(container.querySelector('input')).toHaveFocus();
    userEvent.tab();
    expect(inc).toHaveFocus();
  });

  test('does not emit value update event if the value is outside range / boundary via controls', () => {
    const handleClick = jest.fn();
    render(DefaultQuantity({ min: 4, max: 10, modelValue: 10, 'onUpdate:modelValue': handleClick }));
    fireIncControl();
    expect(handleClick).not.toHaveBeenCalled();
  });

  test('emits value update event on clicking controls', () => {
    const handleClick = jest.fn();
    render(DefaultQuantity({ 'onUpdate:modelValue': handleClick, modelValue: 1 }));
    fireIncControl();
    expect(handleClick).toHaveBeenCalledWith(2);
  });

  test('emits value update event on changing input', () => {
    const handleClick = jest.fn();
    const { container } = render(DefaultQuantity({ 'onUpdate:modelValue': handleClick }));
    const qtyInput = container.querySelector('input') as HTMLInputElement;
    userEvent.type(qtyInput, fieldValue);
    expect(handleClick).toHaveBeenCalledWith(fieldValue);
  });

  test('renders quantity in disabled state', () => {
    const { container } = render(DisabledQuantity());
    expect(container).toMatchSnapshot();
  });

  it('pressing tab should not focus disabled controls', () => {
    render(DisabledQuantity());
    userEvent.tab();
    // Here first control is taken for testing
    expect(screen.getAllByRole('button')[0]).not.toHaveFocus();
  });

  test('does not update value or emit event when in disabled state', () => {
    const handleClick = jest.fn();
    const { container } = render(DefaultQuantity({ disabled: true, 'onUpdate:modelValue': handleClick }));
    fireControls();
    expect(handleClick).not.toHaveBeenCalled();
    const qtyInput = container.querySelector('input') as HTMLInputElement;
    expect(qtyInput.disabled).toBeTruthy();
  });

  test('renders quantity in readonly state', () => {
    const { container } = render(ReadOnlyQuantity());
    expect(container).toMatchSnapshot();
  });

  test('does not update value or emit event when in readonly state', () => {
    const handleClick = jest.fn();
    const { container } = render(DefaultQuantity({ name: fieldName, readonly: true, 'onUpdate:modelValue': handleClick }));
    fireControls();
    expect(handleClick).not.toHaveBeenCalled();
    const qtyInput = container.querySelector('input') as HTMLInputElement;
    expect(qtyInput.readOnly).toBeTruthy();
  });

  test('renders quantity in invalid state', () => {
    const { container } = render(InvalidQuantity());
    expect(container).toMatchSnapshot();
  });
});
