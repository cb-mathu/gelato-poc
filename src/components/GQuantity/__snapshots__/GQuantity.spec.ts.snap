// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GQuantity :  renders quantity in disabled state 1`] = `
<div>
  <div
    class="g-quantity g-quantity--disabled"
    value="5"
  >
    <button
      aria-disabled="true"
      as="button"
      class="g-quantity__change g-quantity__decrement"
      disabled=""
      tabindex="-1"
    >
      <svg
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5 10a1 1 0 011-1h8a1 1 0 110 2H6a1 1 0 01-1-1z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <input
      aria-label="units"
      aria-live="assertive"
      autocomplete="off"
      class="g-quantity__element"
      disabled=""
      id="GQuantity_7"
      invalid="false"
      modelmodifiers="[object Object]"
      modelvalue=""
      name=""
      placeholder=""
      type="number"
    />
    <button
      aria-disabled="true"
      as="button"
      class="g-quantity__change g-quantity__increment"
      disabled=""
      tabindex="-1"
    >
      <svg
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`GQuantity :  renders quantity in invalid state 1`] = `
<div>
  <div
    class="g-quantity g-quantity--invalid"
    value="5"
  >
    <button
      aria-disabled="false"
      as="button"
      class="g-quantity__change g-quantity__decrement"
      tabindex="0"
    >
      <svg
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5 10a1 1 0 011-1h8a1 1 0 110 2H6a1 1 0 01-1-1z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <input
      aria-label="units"
      aria-live="assertive"
      autocomplete="off"
      class="g-quantity__element"
      id="GQuantity_12"
      invalid="true"
      modelmodifiers="[object Object]"
      modelvalue=""
      name=""
      placeholder=""
      type="number"
    />
    <button
      aria-disabled="false"
      as="button"
      class="g-quantity__change g-quantity__increment"
      tabindex="0"
    >
      <svg
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`GQuantity :  renders quantity in readonly state 1`] = `
<div>
  <div
    class="g-quantity g-quantity--readonly"
    value="5"
  >
    <button
      aria-disabled="true"
      as="button"
      class="g-quantity__change g-quantity__decrement"
      disabled=""
      tabindex="-1"
    >
      <svg
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5 10a1 1 0 011-1h8a1 1 0 110 2H6a1 1 0 01-1-1z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <input
      aria-label="units"
      aria-live="assertive"
      autocomplete="off"
      class="g-quantity__element"
      id="GQuantity_10"
      invalid="false"
      modelmodifiers="[object Object]"
      modelvalue=""
      name=""
      placeholder=""
      readonly=""
      type="number"
    />
    <button
      aria-disabled="true"
      as="button"
      class="g-quantity__change g-quantity__increment"
      disabled=""
      tabindex="-1"
    >
      <svg
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`GQuantity :  renders quantity with default args 1`] = `
<div>
  <div
    class="g-quantity"
  >
    <button
      aria-disabled="false"
      as="button"
      class="g-quantity__change g-quantity__decrement"
      tabindex="0"
    >
      <svg
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5 10a1 1 0 011-1h8a1 1 0 110 2H6a1 1 0 01-1-1z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <input
      aria-label="units"
      aria-live="assertive"
      autocomplete="off"
      class="g-quantity__element"
      id="GQuantity_1"
      invalid="false"
      modelmodifiers="[object Object]"
      modelvalue=""
      name=""
      placeholder=""
      type="number"
    />
    <button
      aria-disabled="false"
      as="button"
      class="g-quantity__change g-quantity__increment"
      tabindex="0"
    >
      <svg
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`GQuantity :  renders quantity with native attributes 1`] = `
<div>
  <div
    class="g-quantity"
    max="5"
    min="4"
    step="2"
  >
    <button
      aria-disabled="true"
      as="button"
      class="g-quantity__change g-quantity__decrement"
      disabled=""
      tabindex="-1"
    >
      <svg
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5 10a1 1 0 011-1h8a1 1 0 110 2H6a1 1 0 01-1-1z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <input
      aria-label="units"
      aria-live="assertive"
      autocomplete="off"
      class="g-quantity__element"
      id="GQuantity_2"
      invalid="false"
      max="5"
      min="4"
      modelmodifiers="[object Object]"
      modelvalue=""
      name=""
      placeholder=""
      step="2"
      type="number"
    />
    <button
      aria-disabled="false"
      as="button"
      class="g-quantity__change g-quantity__increment"
      tabindex="0"
    >
      <svg
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;
